src/joint.dia.cell.js

```diff
@@ -201,17 +219,34 @@ joint.dia.Cell = Backbone.Model.extend({
 
             opt = opt || {};
 
-            var z = (this.graph.getFirstCell().get('z') || 0) - 1;
+            var z = this.graph.getCells().reduce(function (minZ, cell) {
+                return Math.min(minZ, cell.get('z'));
+            }, Infinity);
 
-            this.startBatch('to-back');
+            var cells;
 
             if (opt.deep) {
-
-                var cells = this.getEmbeddedCells({ deep: true, breadthFirst: true });
-                cells.reverse().forEach(function(cell) { cell.set('z', z--, opt); });
+                cells = this.getEmbeddedCells({ deep: true, breadthFirst: true });
+                cells.unshift(this);
+            } else {
+                cells = [this];
             }
 
-            this.set('z', z, opt).stopBatch('to-back');
+            var shouldUpdate = cells.reduce(function(shouldUpdate, cell, index) {
```

Comment:
Using `cells.some()`might be more efficient here. (It should be supported from IE9+).

Reply:
done